### PR TITLE
Fix #266 (Focus hijacking)

### LINKED ISSUE #266: Force normal mode
When I open an URL like Messenger.com, I'm automatically in insert mode. ~In fact, any~ Other webpages with a textarea also trigger that behavior. So, I would like to know if it's possible to always force the normal mode when I open an URL (like it should be in VIM) ?

**edit**
It looks like Messenger.com is stealing my focus evertime a switch to an other thread.

### PR DESCRIPTION
**Pull request by glacambre**
This fixes https://github.com/cmcaine/tridactyl/issues/266
This is done by hijacking the page's `HTMLElement.prototype.focus()` function. I am not very satisfied with this solution but it's the only way I found to reliably prevent autofocus, merely using event listeners for focus/focusin events didn't cut it.

We already had a whole discussion about hijacking functions which can be found here: https://github.com/cmcaine/tridactyl/pull/225 . The gist of it is that security currently isn't negatively impacted by hijacking functions, although this might not be true once WebComponents are implemented, hence we need to follow their implementation in Firefox closely and re-evaluate the situation once it lands in Firefox Nightly.

### CODE REVIEW

**Comment by glacambre** [comment]
I just noticed this issue in vim-vixen's repo: https://github.com/ueokande/vim-vixen/issues/333
Basically there's a firefox setting that lets you disable autofocus. Pros to using this instead of merging my PR would be less code in Tridactyl. Cons would be that you have to go through about:config to activate it, that you can't toggle it through pentadactyl, that you can't have a per-domain setting (while this is in theory possible for this PR - once domain-specific settings are implemented) and that the setting could disappear one day (although this is unlikely in my opinion).
In my opinion there are more advantages than disadvantages to merging this PR, but it's always good to know that there are options in case you disagree.

File: src/config.ts [hunk 1]
```diff
@@ -157,6 +157,7 @@ const DEFAULTS = o({
     "homepages": [],
     "hintchars": "hjklasdfgyuiopqwertnmzxcvb",
 	"hintfiltermode": "simple",   // "simple", "vimperator", "vimperator-reflow"
+    "allowautofocus": false,
```
Comment (bovine3dom) [inline src/config.ts]: `false` should be a string, "false". Set cannot set boolean values.

I think it should be "true" by default, as that is the standard behaviour in Firefox (and Vimperator?).

File: src/dom.ts [hunk 2]
```diff
@@ -271,3 +272,54 @@ export function compareElementArea(a: HTMLElement, b: HTMLElement): number {
...
+ *
+ *  Remember to check whether WebComponents change anything security-wise:
+ *  https://developer.mozilla.org/en-US/docs/Web/Web_Components/Custom_Elements
+ *  https://bugzilla.mozilla.org/show_bug.cgi?id=1406825
+ * */
+function onPageFocus(elem: HTMLElement, args: any[]): void {
+    if (isTextEditable(elem)) {
+        LAST_USED_INPUT = elem
+        config.getAsync("allowautofocus").then((allow) => {
+            if (allow)
```
Comment (bovine3dom) [inline src/dom.ts]: Everywhere this is used, it should be changed to a string comparison: `if (allow == "true")`

**Comment by glacambre** [comment]
I implemented the two changes you suggested. Thank you for reviewing my code! :)

**Comment by gilbertw1** [comment]
I haven't had a chance to test out this PR, but I will mention that toggling the `browser.autofocus` setting does not prevent slack from still stealing input on page load.

**Comment by glacambre** [comment]
I realized that `browser.autofocus` only prevents html elements from stealing the focus. This means that pages can still steal focus with JS. Hence we need both `browser.autofocus` and this `.focus()` hijacking PR.
I'll fix the current conflict later today.

**Comment by glacambre** [comment]
Rebased. This still fixes #266. I cant test whether #383 is fixed because I'm getting a 404 when trying to access the page, despite seeing the article being displayed for half a second before disappearing...

**Comment by bovine3dom** [comment]
https://motherboard.vice.com/ gives me 404s too using this PR, but not without this PR, so I think you're breaking the site somehow :(

**Comment by glacambre** [comment]
Removing the last commit fixes this. Now I can confirm that this PR also fixes #383 .

**Pull request merged**